test/irb/helper.rb

```diff
@@ -73,4 +78,109 @@ def without_rdoc(&block)
       }
     end
   end
+
+  class IntegrationTestCase
+    LIB = File.expand_path("../../lib", __dir__)
+    TIMEOUT_SEC = 3
+
+    def setup
+      unless defined?(PTY)
+        omit "Integration tests require PTY."
+      end
+    end
+
+    def run_ruby_file(&block)
+      cmd = [EnvUtil.rubybin, "-I", LIB, @ruby_file.to_path]
+      tmp_dir = Dir.mktmpdir
+
+      @commands = []
+      lines = []
+
+      yield
+
+      PTY.spawn(integration_envs.merge("TERM" => "dumb"), *cmd) do |read, write, pid|
+        Timeout.timeout(TIMEOUT_SEC) do
+          while line = safe_gets(read)
+            lines << line
+
+            # means the breakpoint is triggered
+            if line.match?(/binding\.irb/)
+              while command = @commands.shift
+                write.puts(command)
+              end
+            end
+          end
+        end
+      ensure
+        read.close
+        write.close
+        kill_safely(pid)
+      end
+
+      lines.join
+    rescue Timeout::Error
+      message = <<~MSG
+      Test timedout.
+
+      #{'=' * 30} OUTPUT #{'=' * 30}
+        #{lines.map { |l| "  #{l}" }.join}
+      #{'=' * 27} END OF OUTPUT #{'=' * 27}
+      MSG
+      assert_block(message) { false }
+    ensure
+      File.unlink(@ruby_file) if @ruby_file
+      FileUtils.remove_entry tmp_dir
+    end
+
+    # read.gets could raise exceptions on some platforms
+    # https://github.com/ruby/ruby/blob/master/ext/pty/pty.c#L729-L736
+    def safe_gets(read)
+      read.gets
+    rescue Errno::EIO
+      nil
+    end
+
+    def kill_safely pid
+      return if wait_pid pid, TIMEOUT_SEC
+
+      Process.kill :TERM, pid
+      return if wait_pid pid, 0.2
+
+      Process.kill :KILL, pid
+      Process.waitpid(pid)
+    rescue Errno::EPERM, Errno::ESRCH
+    end
+
+    def wait_pid pid, sec
+      total_sec = 0.0
+      wait_sec = 0.001 # 1ms
+
+      while total_sec < sec
+        if Process.waitpid(pid, Process::WNOHANG) == pid
+          return true
+        end
+        sleep wait_sec
+        total_sec += wait_sec
+        wait_sec *= 2
+      end
+
+      false
+    rescue Errno::ECHILD
+      true
+    end
+
+    def type(command)
+      @commands << command
+    end
+
+    def write_ruby(program)
+      @ruby_file = Tempfile.create(%w{irb- .rb})
+      @ruby_file.write(program)
+      @ruby_file.close
+    end
```

Comment:
Just an idea

I know that this code is not changed from `test_debug_cmd.rb`, but could it be like this?
```ruby
ruby_code = <<~RUBY
  foo
  binding.irb
  bar
RUBY
run_ruby_file ruby_code do end
```

This helper will going to be used in several test, so it's a good opportunity to change it.

`write_ruby` was introduced in https://github.com/ruby/irb/commit/976100c1c26808b9d8c58f72442dbc01c13a790f.
In that point, defining `write_ruby` like another existing method `write_irbrc` makes sense because it's optional. Not always used.
Now, write_ruby is always used before run_ruby_file. I think passing as an argument is more simple.


Although, my idea has a weak point. Must think how to do something like
```
assert_match(/Object#foo at #{@ruby_file.to_path}/, output)
```

Reply:
I agree with the direction, but I feel we can design the API better after seeing other use cases. Can we revisit this after we applied integration for #659's fix?